CMakeLists.txt

```diff
@@ -0,0 +1,198 @@
+cmake_minimum_required(VERSION 3.14)
+project(mold VERSION 1.3.0)
+
+include(CMakeDependentOption)
+
+# FIXME: this is for parity with the makefiles that install directly to <prefix>/lib
+# NOTE: defining this before including GNUInstallDirs makes its platform detection a noop
+set(CMAKE_INSTALL_LIBDIR lib CACHE PATH "")
+include(GNUInstallDirs)
+
+# Mold build options
+cmake_dependent_option(MOLD_USE_MIMALLOC "Use mimalloc" ON
+    "NOT APPLE;NOT ANDROID" OFF)
+cmake_dependent_option(MOLD_USE_SYSTEM_MIMALLOC "Use system or vendored mimalloc" ON
+    MOLD_USE_MIMALLOC OFF)
+option(MOLD_USE_SYSTEM_TBB "Use system or vendored TBB" ON)
+
+# Setup mimalloc
+if(MOLD_USE_MIMALLOC)
+    if(MOLD_USE_SYSTEM_MIMALLOC)
+        find_package(mimalloc REQUIRED)
+        target_compile_definitions(mimalloc INTERFACE USE_SYSTEM_MIMALLOC)
+    else()
+        function(mold_add_mimalloc)
+            option(MI_BUILD_TESTS "Build test executables" OFF)
+            add_subdirectory(third-party/mimalloc EXCLUDE_FROM_ALL)
+            target_compile_definitions(mimalloc PRIVATE MI_USE_ENVIRON=0)
+        endfunction()
+
+        mold_add_mimalloc()
+    endif()
+endif()
+
+# Setup TBB
+if(MOLD_USE_SYSTEM_TBB)
+    find_package(TBB REQUIRED)
+else()
+    function(mold_add_tbb)
+        set(BUILD_SHARED_LIBS OFF)
+        set(TBB_TEST OFF CACHE INTERNAL "")
+        set(TBB_STRICT OFF CACHE INTERNAL "")
+        add_subdirectory(third-party/tbb EXCLUDE_FROM_ALL)
+        target_compile_definitions(tbb PRIVATE __TBB_DYNAMIC_LOAD_ENABLED=0)
+    endfunction()
+
+    mold_add_tbb()
+endif()
+
+add_library(mold-wrapper SHARED)
+
+# Remove the default `lib` prefix
+set_target_properties(mold-wrapper PROPERTIES PREFIX "")
+target_link_libraries(mold-wrapper PRIVATE ${CMAKE_DL_LIBS})
+target_sources(mold-wrapper PRIVATE elf/mold-wrapper.c)
+
+add_executable(mold)
+target_compile_features(mold PRIVATE cxx_std_20)
+target_include_directories(mold SYSTEM PRIVATE third-party)
```

Comment:
This should probably be brought into the mold executable build via target_link_libraries, and not by adding the include directory to mold itself.

Reply:
Now that you mention it... the includes in this directory might be picked up before the system dependencies because of this. There should be a separate PR that fixes includes:

```diff
diff --git a/demangle.cc b/demangle.cc
--- demangle.cc
+++ demangle.cc
@@ -1,9 +1,9 @@
 #include "mold.h"
 
 #include <cstdlib>
 #include <cxxabi.h>
-#include <rust-demangle/rust-demangle.h>
+#include <rust-demangle.h>
 
 namespace mold {
 
 std::string_view demangle(std::string_view name) {
diff --git a/Makefile b/Makefile
--- Makefile
+++ Makefile
@@ -44,9 +44,10 @@
 CFLAGS = -O2
 CXXFLAGS = -O2
 
 MOLD_CXXFLAGS := -std=c++20 -fno-exceptions -fno-unwind-tables \
-                 -fno-asynchronous-unwind-tables -Ithird-party \
+                 -fno-asynchronous-unwind-tables \
+                 -Ithird-party/rust-demangle -Ithird-party/xxhash \
                  -DMOLD_VERSION=\"$(VERSION)\" -DLIBDIR="\"$(LIBDIR)\""
 
 MOLD_LDFLAGS := -pthread -lz -lm -ldl
 
diff --git a/mold.h b/mold.h
--- mold.h
+++ mold.h
@@ -26,9 +26,9 @@
 #include <unistd.h>
 #include <vector>
 
 #define XXH_INLINE_ALL 1
-#include <xxhash/xxhash.h>
+#include <xxhash.h>
 
 #ifdef NDEBUG
 #  define unreachable() __builtin_unreachable()
 #else
```

Reply:
Submitted the above patch as #601